The ECDH and X25519 deriveBits returns an empty string when 'length' is 0

javifernandez opened the pull request:
#### 0b46039641c2700f55501d4121d825aa86f22b98
<pre>
The ECDH and X25519 deriveBits returns an empty string when &apos;length&apos; is 0
<a href="https://bugs.webkit.org/show_bug.cgi?id=276916">https://bugs.webkit.org/show_bug.cgi?id=276916</a>

Reviewed by Matthew Finkel.

The WebCrypto API spec&apos;s draft states that the ECDH&apos;s deriveBits should
handle a zero length as any regular number, only throwing an exception
in case of &apos;null&apos;. The same is stated in the Secure Curves specification
draft for the X25519 algorithm.

We were not supporting &apos;null&apos; value before, but since r281240 the &apos;length&apos;
parameter is defined as optional, with &apos;null&apos; as default value. Hence
in case of a zero length the derived bits are truncated so that the
operation returns an empty string.

* LayoutTests/crypto/subtle/ecdh-derive-bits-length-limits-expected.txt:
* LayoutTests/crypto/subtle/ecdh-derive-bits-length-limits.html:
* LayoutTests/imported/w3c/web-platform-tests/WebCryptoAPI/derive_bits_keys/derived_bits_length.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/WebCryptoAPI/derive_bits_keys/derived_bits_length.https.any.worker-expected.txt:
* Source/WebCore/crypto/algorithms/CryptoAlgorithmECDH.cpp:
(WebCore::CryptoAlgorithmECDH::deriveBits):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmX25519.cpp:
(WebCore::CryptoAlgorithmX25519::deriveBits):

Canonical link: <a href="https://commits.webkit.org/285383@main">https://commits.webkit.org/285383@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7c8d375ec46eaf92cad5db4d868055cf97b307e6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71617 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51030 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24391 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/75729 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22822 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/73732 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58831 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22642 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56568 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15048 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74683 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46297 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61693 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37017 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42960 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19160 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21163 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64864 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19523 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77447 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15851 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18702 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/64282 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15894 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61726 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64279 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16002 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12426 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6067 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46830 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/1609 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47901 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49185 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47643 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nmahendru commented on Source/WebCore/crypto/algorithms/CryptoAlgorithmECDH.cpp:
```diff
@@ -94,7 +94,7 @@ void CryptoAlgorithmECDH::deriveBits(const CryptoAlgorithmParameters& parameters
             return;
         }
-        if (!length || !(*length)) {
+        if (!length) {
```
This technically does work. But I think it would be wise to check for null as the first thing in the function ::deriveBits and bail early.
For Zero, I think we can check right before this lambda and don't even get into doing scalar multiplication and just return a zero length result.

javifernandez replied:
I agree on the optimization for the zero length, which would even allow us to avoid the call to the platformDeriveBits key. 

However I'm not so sure about the early bail in case of null; since the change to define length as optional, null means that we should return the full derivedKey. Hence, I don't think we save much to avoid doing the check inside the lambda.

javifernandez replied:
Also, not a strong opinion, but I think it's better to do the length = 0 after the rest of the basic checks (eg. ec parameters, key type, algorithm name). Otherwise we would hide failures in those checks when passing 0 as length.

nmahendru replied:
> Otherwise we would hide failures in those checks when passing 0 as length.


It's true vice versa. If we were adding those other checks in this PR instead of zero length, do you think the same argument applies ? I might be missing stuff in the spec. I had a quick look.

nmahendru replied:
>  optional, null means that we should return the full derivedKey. 

I did not know that. Then I guess we need to derive the key.

javifernandez replied:
> > Otherwise we would hide failures in those checks when passing 0 as length.
> 
> It's true vice versa. If we were adding those other checks in this PR instead of zero length, do you think the same argument applies ? I might be missing stuff in the spec. I had a quick look.

Very true; my preference is based on the fact that the 'length' behavior is based on the specific algorithm ; for instance, PBKDF2 and HKDF would throw in case of 0, and we are aiming for ECDH and X25519 to throw on any truncation. So in my mind it makes more sense to ensure first that the algorithm name, key type and usages are validates and only then check the specific values of the 'length' parameter. 

But as I said, I don't have a strong opinion on this, because you have a valid point and perhaps it's more efficient to early return based on an "cheap" check of the length's value, avoiding the other checks.

nmahendru reviewed (changes requested):
It does work technically but with the spec change the code can be easily made a little better at failing or returning fast.

nmahendru commented on Source/WebCore/crypto/algorithms/CryptoAlgorithmX25519.cpp:
```diff
@@ -87,12 +87,20 @@ void CryptoAlgorithmX25519::deriveBits(const CryptoAlgorithmParameters& paramete
+    // https://github.com/WICG/webcrypto-secure-curves/pull/29
+    if (length && !(*length)) {
+        // Avoid executing the key-derivation, since we are going to return an empty string.
+        callback(Vector<uint8_t>());
```
Assuming there is no style error in this.
```suggestion
        callback( { } );
```

javifernandez replied:
Done

nmahendru commented on Source/WebCore/crypto/algorithms/CryptoAlgorithmX25519.cpp:
```diff
@@ -87,12 +87,20 @@ void CryptoAlgorithmX25519::deriveBits(const CryptoAlgorithmParameters& paramete
 
+    // Return an empty string doesn't make much sense, but truncating either at all.
+    // https://github.com/WICG/webcrypto-secure-curves/pull/29
+    if (length && !(*length)) {
```
super nit:
```suggestion
    if (!length.value_or(0)) {
```

javifernandez replied:
I think this is not correct; we would return an empty string in case "null".  The [spec](https://w3c.github.io/webcrypto/#ecdh-operations) states:

> If length is null:
    Return secret

nmahendru replied:
ohh. right.. then what you have is perfect.
To re-state, Truncate only if length is explicitly provided!

nmahendru reviewed (approved):
Thanks for all the work on this PR and the spec!

sysrqb commented on LayoutTests/crypto/subtle/ecdh-derive-bits-length-limits.html:
```diff
@@ -85,7 +85,7 @@
         // P-256
         return Promise.all([
-            deriveBits(P256, 0, 256),
+            deriveBits(P256, 0, 0),
```
Is this test cases for 256 bits, and similar below, not valuable anymore? How does that relate to the 0-length check?

javifernandez replied:
> Is this test cases for 256 bits, and similar below, not valuable anymore? How does that relate to the 0-length check?

The test case checked that when passing 0 as 'length', the result was the full string; now we should return an empty string instead.

sysrqb reviewed (approved):
Changes lgtm.

javifernandez committed:
The ECDH and X25519 deriveBits returns an empty string when 'length' is 0
https://bugs.webkit.org/show_bug.cgi?id=276916

Reviewed by Matthew Finkel.

The WebCrypto API spec's draft states that the ECDH's deriveBits should
handle a zero length as any regular number, only throwing an exception
in case of 'null'. The same is stated in the Secure Curves specification
draft for the X25519 algorithm.

We were not supporting 'null' value before, but since r281240 the 'length'
parameter is defined as optional, with 'null' as default value. Hence
in case of a zero length the derived bits are truncated so that the
operation returns an empty string.

* LayoutTests/crypto/subtle/ecdh-derive-bits-length-limits-expected.txt:
* LayoutTests/crypto/subtle/ecdh-derive-bits-length-limits.html:
* LayoutTests/imported/w3c/web-platform-tests/WebCryptoAPI/derive_bits_keys/derived_bits_length.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/WebCryptoAPI/derive_bits_keys/derived_bits_length.https.any.worker-expected.txt:
* Source/WebCore/crypto/algorithms/CryptoAlgorithmECDH.cpp:
(WebCore::CryptoAlgorithmECDH::deriveBits):
* Source/WebCore/crypto/algorithms/CryptoAlgorithmX25519.cpp:
(WebCore::CryptoAlgorithmX25519::deriveBits):

Canonical link: https://commits.webkit.org/285383@main
